File: gene_network_modeling/oscillator_ssa.py
import numpy as np
import matplotlib
matplotlib.use('tkagg')
import matplotlib.pyplot as plt
import random


G1 = [0]
G2 = [0]
G3 = [0]
t = [0]

tend = 1000

k_1 = 2
gamma_1 = 0.1
k_2 = 2
gamma_2 = 0.1
k_3 = 2
gamma_3 = 0.1
n = 9
c = 1


while t[-1] < tend:

        current_G1 = G1[-1]
        current_G2 = G2[-1]
        current_G3 = G3[-1]

        rates = [(c**n / (c**n + current_G3**n)) * k_1, gamma_1 * current_G1, \
        (current_G1**n / (c**n + current_G1**n)) * k_2, gamma_2 * current_G2, \
        (current_G2**n / (c**n + current_G2**n)) * k_3, gamma_3 * current_G3]

        rate_sum = sum(rates)

        tau = np.random.exponential(scale=1/rate_sum)

        t.append(t[-1] + tau)

        rand = random.uniform(0,1)


        # G1 production event
        if rand * rate_sum <= rates[0]:
                G1.append(G1[-1] + 1)
                G2.append(G2[-1])
                G3.append(G3[-1])

        # G1 decay event
        elif rand * rate_sum > rates[0] and rand * rate_sum <= sum(rates[:2]):
                G1.append(G1[-1] - 1)
                G2.append(G2[-1])
                G3.append(G3[-1])

        # G2 production event
        elif rand * rate_sum > sum(rates[:2]) and rand * rate_sum <= sum(rates[:3]):

                G1.append(G1[-1])
                G2.append(G2[-1] + 1)
                G3.append(G3[-1])

        # G2 decay event
        elif rand * rate_sum > sum(rates[:3]) and rand * rate_sum <= sum(rates[:4]):

                G1.append(G1[-1])
                G2.append(G2[-1] - 1)
                G3.append(G3[-1])

        # G3 production event
        elif rand * rate_sum > sum(rates[:4]) and rand * rate_sum <= sum(rates[:5]):

                G1.append(G1[-1])
                G2.append(G2[-1])
                G3.append(G3[-1] + 1)

        # G3 decay event
        elif rand * rate_sum > sum(rates[:5]) and rand * rate_sum <= sum(rates[:6]):
                G1.append(G1[-1])
                G2.append(G2[-1])
                G3.append(G3[-1] - 1)



f, (ax1, ax2, ax3) = plt.subplots(3, sharex=True, sharey=False)
line1, = ax1.plot(t , G1, color="b",label="G1")
line2, = ax2.plot(t , G2, color="r",label="G2")
line3, = ax3.plot(t , G3, color="g",label="G3")
ax1.set_ylabel('Number')
ax1.set_xlabel('Time')
ax1.legend(handles=[line1,line2,line3])
plt.show()

File: gene_network_modeling/oscillator_ssa.m


close all
clear all
clc



tend = 1000;

k_1 = 2;
gamma_1 = 0.1;
k_2 = 2;
gamma_2 = 0.1;
k_3 = 2;
gamma_3 = 0.1;
n = 9;
c = 1;

t(1) = 0;
G1(1) = 0;
G2(1) = 0;
G3(1) = 0;


i = 1;


while t(end) < tend
    
    rates = zeros(1,6);
    
    rates(1) = (c^n / (c^n + G3(i)^n)) * k_1;
    rates(2) = gamma_1 * G1(i);
    
    rates(3) = (G1(i)^n / (c^n + G2(i)^n)) * k_2;
    rates(4) = gamma_2 * G2(i);
    
    rates(5) = (G2(i)^n / (c^n + G2(i)^n)) * k_3;
    rates(6) = gamma_3 * G3(i);
    
    
    rate_sum = sum(rates);
    
    
    r1 = rand(1);
    tau = (1/rate_sum) * log(1/r1);
    
    
    t(i+1) = t(i) + tau;
    
    
    r2 = rand(1);
    
    if r2*rate_sum <= rates(1)
        G1(i+1) = G1(i) + 1;
        G2(i+1) = G2(i);
        G3(i+1) = G3(i);
      
    elseif r2 * rate_sum > rates(1) && r2 * rate_sum <= rates(1) + rates(2)
        G1(i+1) = G1(i) - 1;
        G2(i+1) = G2(i);
        G3(i+1) = G3(i);            
        
    elseif r2 * rate_sum > rates(1) + rates(2) && r2 * rate_sum <= rates(1) + rates(2) + rates(3)
            G1(i+1) = G1(i);
            G2(i+1) = G2(i) + 1;
            G3(i+1) = G3(i);            
       
    elseif r2 * rate_sum > rates(1) + rates(2) + rates(3) && r2 * rate_sum <= rates(1) + rates(2) + rates(3) + rates(4)
        G1(i+1) = G1(i);
        G2(i+1) = G2(i) - 1;
        G3(i+1) = G3(i);            
 
        
    elseif r2 * rate_sum > rates(1) + rates(2) + rates(3) + rates(4) && r2 * rate_sum <= rates(1) + rates(2) + rates(3) + rates(4) + rates(5)
        G1(i+1) = G1(i);
        G2(i+1) = G2(i);        
        G3(i+1) = G3(i) + 1;
        
        
    elseif r2 * rate_sum > rates(1) + rates(2) + rates(3) + rates(4) + rates(5) && r2 * rate_sum <= rates(1) + rates(2) + rates(3) + rates(4) + rates(5) + rates(6)
        G1(i+1) = G1(i);
        G2(i+1) = G2(i);             
        G3(i+1) = G3(i) - 1;
        
        
        
    end
    
   
    i = i + 1;
    
    
    
end


subplot(3,1,1);
plot(t,G1)

subplot(3,1,2);
plot(t,G2)

subplot(3,1,3);
plot(t,G3)






